doc/plugins/authoring.rst

```diff
@@ -0,0 +1,221 @@
+Authoring a Plugin
+====
+
+NixOps plugins extend NixOps' core to support additional hosting
+providers and resource types.
+
+Some example plugins include:
+
+- https://github.com/NixOS/nixops-aws
+- https://github.com/NixOS/nixops-hetzner
+- https://github.com/nix-community/nixops-vbox
+- https://github.com/nix-community/nixops-libvirtd
+- https://github.com/nix-community/nixops-datadog
+
+This guide is light on the details, and intends to describe just the
+supported hooks and integration process.
+
+Packaging with Poetry and poetry2nix
+====
+
+NixOps and its plugins are packaged as standard Python applications.
+Most packages will use `Poetry <https://python-poetry.org>`_ and
+`poetry2nix <https://github.com/nix-community/poetry2nix>`_ for
+packaging with Nix.
+
+Note: NixOps is formatted with ``black`` and strictly typechecked with
+``mypy``. Your project should follow these guidelines as well, and use
+a mypy configuration at least as strict as the NixOps mypy
+configuration.
+
+First, create a ``pyproject.toml`` to describe your project. This is
+intsead of a ``setup.py``, and using both may cause confusing build
+errors. Only use a ``pyproject.toml``::
+
+  [tool.poetry]
+  name = "nixops_neatcloud"
+  version = "1.0"
+  description = "NixOps plugin for NeatCloud"
+  authors = ["Your Name <your.name@example.com>"]
+  license = "MIT"
+  include = [ "nixops_neatcloud/nix/*.nix" ]
+
+  [tool.poetry.dependencies]
+  python = "^3.7"
+  nixops = {git = "https://github.com/NixOS/nixops.git", rev = "master"}
+
+  [tool.poetry.plugins."nixops"]
+  neatcloud = "nixops_neatcloud.plugin"
+
+  [build-system]
+  requires = ["poetry>=0.12"]
+  build-backend = "poetry.masonry.api"
+
+Now create your first ``poetry.lock`` file with ``poetry lock``::
+
+  nixops_neatcloud$ nix-shell -p poetry
+  [nix-shell:nixops_neatcloud]$ poetry lock
+  Creating virtualenv nixops_neatcloud-FrXThxiS-py3.7 in ~/.cache/pypoetry/virtualenvs
+  Updating dependencies
+  Resolving dependencies... (2.1s)
+
+  Writing lock file
+
+Exit the Nix shell, and create the supporting Nix files.
+
+Create a ``default.nix``::
+
+  { nixpkgs ? <nixpkgs>
+  , pkgs ? import nixpkgs {}
+  }:
+  let
+    overrides = import ./overrides.nix { inherit pkgs; };
+  in pkgs.poetry2nix.mkPoetryApplication {
+    src = pkgs.lib.cleanSource ./.;
```

Comment:
I think we should encourage using `projectDir` instead of the separate params.
This will apply the default gitignore/pycache source filters and "does the right thing".

Reply:
I agree, but at the time of writing, the 20.03 branch doesn't support it. As soon as the channel contains it, I'll update the docs